samcli/commands/build/build_context.py

```diff
@@ -0,0 +1,94 @@
+
+import os
+import shutil
+import yaml
+
+from samcli.yamlhelper import yaml_parse
+from samcli.commands.local.lib.sam_function_provider import SamFunctionProvider
+
+
+class BuildContext(object):
+
+    def __init__(self, template_file, source_root, build_dir, clean=False, use_container=False):
+        self._template_file = template_file
+        self._source_root = source_root
+        self._build_dir = build_dir
+        self._clean = clean
+        self._use_container = use_container
+
+        self._function_provider = None
+        self._template_dict = None
+        self._app_builder = None
+
+    def __enter__(self):
+        self._template_dict = self._get_template_data(self._template_file)
+        self._function_provider = SamFunctionProvider(self._template_dict)
+
+        self._build_dir = self._setup_build_dir()
+
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @property
+    def function_provider(self):
+        return self._function_provider
+
+    @property
+    def template_dict(self):
+        return self._template_dict
+
+    @property
+    def build_dir(self):
+        return self._build_dir
+
+    @property
+    def source_root(self):
+        return self._source_root
+
+    @property
+    def use_container(self):
+        return self._use_container
+
+    @property
+    def output_template_path(self):
+        return os.path.join(self._build_dir, "built-template.yaml")
+
+    def _setup_build_dir(self):
+
+        # Get absolute path
+        self._build_dir = os.path.abspath(self._build_dir)
+
+        if not os.path.exists(self._build_dir):
+            # TODO: What permissions should I apply to this directory?
```

Comment:
I was going to ask this question. Its possible that this will need execute permissions within the files and the current user should have read/write (I would say). I don't think group or global permissions are needed at all for the build directory. 700?

On the same note, how do we ensure that the files we write or produce in this directory have correctly scoped permissions? Is that handled in the aws_lambda_builders module?

Reply:
This directory is not zipped and passed to Lambda, instead it is only the artifacts that are passed. By default artifacts get written with 755 on directory and 644 on files. But in future we might apply a build action that just verifies this